packages/dht/src/connection/ConnectionManager.ts

```diff
@@ -0,0 +1,384 @@
+}
+
+const DEFAULT_DISCONNECTION_TIMEOUT = 10000
+
+const logger = new Logger(module)
+
+export class ConnectionManager extends EventEmitter implements ITransport {
+    public PROTOCOL_VERSION = '1.0'
+    private CONNECTION_MANAGER_APP_ID = 'connectionmanager'
+    private stopped = false
+    private started = false
+
+    private ownPeerDescriptor: PeerDescriptor | null = null
+    private connections: { [peerId: string]: IConnection } = {}
+
+    private disconnectionTimeouts: { [peerId: string]: NodeJS.Timeout } = {}
+    private webSocketConnector: WebSocketConnector | null
+    private webrtcConnector: WebRtcConnector | null
+
+    private webSocketServer: WebSocketServer | null
+
+    constructor(private config: ConnectionManagerConfig) {
+        super()
+        if (config.webSocketPort) {
+            this.webSocketServer = new WebSocketServer()
+        } else {
+            this.webSocketServer = null
+        }
+        this.webSocketConnector = this.createWsConnector()
+        this.webrtcConnector = this.createWebRtcConnector()
+    }
+
+    private async handleIncomingConnectivityRequest(connection: IConnection, connectivityRequest: ConnectivityRequestMessage) {
+        if (!this.started || this.stopped) {
+            return
+        }
+        let outgoingConnection: IConnection | null = null
+        let connectivityResponseMessage: ConnectivityResponseMessage | null = null
+        try {
+            outgoingConnection = await this.webSocketConnector!.connectAsync({
+                host: (connection as ServerWebSocket).remoteAddress,
+                port: connectivityRequest.port, timeoutMs: 1000
+            })
+        }
+        catch (e) {
+            logger.trace("Connectivity test produced negative result, communicating reply to the requester")
+            logger.debug(e)
+
+            connectivityResponseMessage = {
+                openInternet: false,
+                ip: (connection as ServerWebSocket).remoteAddress,
+                natType: 'unknown'
+            }
+        }
+
+        if (outgoingConnection) {
+            outgoingConnection.close()
+
+            logger.trace("Connectivity test produced positive result, communicating reply to the requester")
+
+            connectivityResponseMessage = {
+                openInternet: true,
+                ip: (connection as ServerWebSocket).remoteAddress,
+                natType: 'open_internet',
+                websocket: { ip: (connection as ServerWebSocket).remoteAddress, port: connectivityRequest.port }
+            }
+        }
+
+        const msg: Message = {
+            appId: this.CONNECTION_MANAGER_APP_ID,
+            messageType: MessageType.CONNECTIVITY_RESPONSE, messageId: v4(),
+            body: ConnectivityResponseMessage.toBinary(connectivityResponseMessage!)
+        }
+        connection.send(Message.toBinary(msg))
+    }
+
+    private async sendConnectivityRequest(): Promise<ConnectivityResponseMessage> {
```

Comment:
Something like this could maybe make the flow easier to read:
```
    private async sendConnectivityRequest(): Promise<ConnectivityResponseMessage> {
        const entryPoint = this.config.entryPoints![0]

        let outgoingConnection: IConnection | null = null

        try {
            outgoingConnection = await this.webSocketConnector!.connectAsync({
                host: entryPoint.websocket?.ip, port: entryPoint.websocket?.port, timeoutMs: 1000
            })
        }
        catch (e) {
            logger.error("Failed to connect to the entrypoints")
            throw new Error('Failed to connect to the entrypoints')
        }

        // prepare for receiving a connectivity reply
        const responsePromise = waitForEvent(outgoingConnection, ConnectionEvents.DATA)

        // send connectivity request
        const connectivityRequestMessage: ConnectivityRequestMessage = { port: this.config.webSocketPort! }
        const msg: Message = {
            appId: this.CONNECTION_MANAGER_APP_ID,
            messageType: MessageType.CONNECTIVITY_REQUEST, messageId: 'xyz',
            body: ConnectivityRequestMessage.toBinary(connectivityRequestMessage)
        }

        outgoingConnection.once(ConnectionEvents.ERROR, () => {
            logger.trace('clientsocket error')
        })

        logger.trace('trying to send connectivity request')
        outgoingConnection.send(Message.toBinary(msg))
        logger.trace('connectivity request sent: ' + JSON.stringify(Message.toJson(msg)))

        const bytes = await responsePromise[0]
        const message: Message = Message.fromBinary(bytes)
        return ConnectivityResponseMessage.fromBinary(message.body)
    }
```

Here also the complier knows the execution path and knows that `outgoingConnection` is always available in line 83:

Reply:
Worth trying out, could make code a tighter. Add enough timeout to `waitForEvent`

Reply:
Moved all connectivity checking code to ConnectivityChecker.ts, sendConnectivityRequest() has been totally rewritten.